### DISCUSSION TITLE
New guy has trouble installing SCIRun 5

### DISCUSSION THREAD

**Discussion by kimjohn1** [SCIInstitute/SCIRun#2232] [Q&A] [answered]
Greetings,

    I am attempting to install SCIRun 5 for the first time and can't seem to get past cmake ../Superbuild.  I get an error message:
CMake Error at Superbuild.cmake:133 (MESSAGE):
  Set Qt5_PATH to directory where Qt 5 is installed (containing lib and bin
  subdirectories) or set BUILD_HEADLESS to ON.

    I have set the path for Qt5 that I found in Superbuild.cmake, with:
SET(Qt5_PATH "/usr/lib/x86_64-linux-gnu/qt5/" CACHE PATH "Path to directory where Qt 5 is installed. Directory should contain lib and bin subdirectories.")

    I have also tried setting BUILD_HEADLESS to ON by changing the option found in Superbuild.cmake to:
OPTION(BUILD_HEADLESS "Build SCIRun without GUI." ON)

Neither change had any effect.

    I am operating with Linux Mint 20.04, and note that I have recently been using Paraview, which uses Qt5, and which works.  When I query the system with 
qmake --version, I get:
QMake version 3.1
Using Qt version 5.9.7 in /home/kj/anaconda3/lib

    When I query the system with whereis qt5, I get:
qt5: /usr/lib/x86_64-linux-gnu/qt5 /usr/lib/qt5 /usr/share/qt5

    When I query the system with whereis Qt5, I get nothing:
Qt5:

    I find Qt5 named folders in /home/kj/anaconda3/lib/cmake
    I find a single qt5 folder in /usr/lib/x86_64-linux-gnu/
    None of the qt5 or Qt5 folders contain "lib and bin subdirectories".
    I will take the system down, re-format the main drive and and re-install the OS tomorrow to give me a clean slate and start again.  Linux Mint includes Qt5 and gcc 9 as part of the OS install, so I will install cmake per the SCIRun 5 Build Requirements, All Platforms, found at:
http://sciinstitute.github.io/SCIRun/build.html

    The SCIRun 5 Build instruction does not provide a download link, nor file identification for a Linux install, nor specify where to locate the un-packed source file, so for my install I will retrieve the source code by going to:
https://github.com/SCIInstitute/SCIRun/releases/tag/v5.0-beta.Y
and clicking on the link for Source code (tar.gz).
    I will unpack the compressed folder and move the unpacked folder, SCIRun-5.0-beta.Y, to my home directory (/home/kj), then attempt to compile per the Compiling SCIRun instructions:
cd bin
cmake ../Superbuild

    After a successful compile, I will change directory back to SCIRun-5.0-beta.Y and run the build script:
./build.sh
    Are there glaring errors in my first attempt to install SCIRun 5, or in my planned course of action, that I can correct and make the installation more likely to be successful?
    Sorry to be so verbose.  As a closing question, can you tell me if the section titled, "Building SCIRun with Qt 5" is an alternative to running the build.sh script, and if so, what would be the advantage there?

Kim

**Comment by dcwhite** [comment]
Hi,
I'm the main developer. I got your email yesterday from the email list. 

First, there's no Linux installer. One must build from source. You can download a zip from the release page, or from the github home page (master branch), or preferably, clone the repo with git so you can get updates more easily (see below).

If you noticed on the build instruction page, Qt 5.13 or later is required, and you're trying 5.9. I suggest installing Qt 5.15.2 to a user directory and pointing to that with cmake. You don't need to edit the cmake files themselves, just run cmake and fill in the directory in the UI. Alternatively, pass it to build.sh along with `-DQt5_PATH=[the right path here]`.

You might have build failures however--you'll need to checkout a branch that's being tested right now where I have updated some additional library dependencies. So after cloning, you can run `git checkout boost172` and rebuild.

And you did find the latest doc on how to build, but it's a bit lacking. We're actively buffing up a lot of areas of the documentation website so we'll get to the build page very soon, based on your experience. SCIRun isn't the easiest build, but the instructions should be clearer. Thanks for being so persistent! 

Let me know here if you have more questions, the more the better!
Dan

**Reply by tpat95** [reply]
I think setting the flag -DQt5_PATH will do the trick for getting through the first build error. I tried setting that flag in the same position of Superbuild but that did not work. Alternatively, you can set this flag when calling `ccmake ../Superbuild`.

**Reply by kimjohn1** [reply]
Thank you both for the excellent and patient replies.   have cleaned the hard drive and installed Linux Mint 20.10, then installed prerequisites, including Qt5 version 5.15.2, which was installed in folder Qt in my home directory.  I've attached a list of the steps I took, organized in an install guide sequence.  Compile, executed from the SCIRun-5.0-beta.Y/bin folder, makes it past the 'can't find Qt5 anywhere' stage, using:
````
cmake -DQt5_PATH=/home/kj/Qt/5.15.2/gcc_64 ../Superbuild
````

Using ccmake -DQt5_PATH=/home/kj/Qt/5.15.2/gcc_64 ../Superbuild also works and presents a little asci based GUI with which I am unfamiliar, and couldn't do much with except select q to make it go away.
Compile still fails, but with a new error:
````
CMake Error at /home/kj/Qt/5.15.2/gcc_64/lib/cmake/Qt5Gui/Qt5GuiConfigExtras.cmake:9 (message):
  Failed to find "GL/gl.h" in "/usr/include/libdrm".
````

If you think this new failure might be fixed by the boost172 branch that you mention, send me a little more detail about how I implement using that, and I'll give it a whirl.

Thanks again for the help.  I spent from 1:00 to 10:00 getting the new OS installed, prerequisites installed and the cmake flag figured out.  The majority of the time was spent swearing my way through the Qt5 install.  SCIRun 5 is not hard to install.  Qt5 is hard to install.  I experienced a similar circumstance when installing Singularity.  That requires installation of Go, which seemed to be endlessly difficult for me to install.  I spent a day writing an installation script for a complete Singularity installation, including Go, and have used that ever since.  Works like a charm, but I don't think that would be possible with Qt5. 

Let me know if you have any ideas about fixing or working around that compile error.

Kim
[SCIRun_Install_Guide.pdf](https://github.com/SCIInstitute/SCIRun/files/6014229/SCIRun_Install_Guide.pdf)

**Comment by kimjohn1** [comment]
A couple more data points:
1. After manually installing Qt5, I can set the path with
````
export PATH=/home/kj/Qt/5.15.2/gcc_64:$PATH
````
No need for the compile flag after the path is set.  I just use
````
cmake ../Superbuild
````
2. spack can be used to install Qt5 version 5.15.2
````
 	git clone https://github.com/spack/spack
	spack install qt@5.15.2
	spack load qt
        export PATH=/home/kj/spack/opt/spack/linux-ubuntu20.10-westmere/gcc-10.2.0/qt-5.15.2-up6rrgb42pwybyn2bntm3q6itsagcdc2:$PATH
````
3. spack uses curl, so I added curl to the dependencies that get installed in my install guide notes.  The spack installation is easy to do and works like a champ, but in typical spack fashion, it takes an extraordinarily long time.  The install step took 2.5 hours when I ran it today.

4. I installed Ubuntu 20.10 on the lab computer and went through the SCIRun install.  Result was similar to the experience with Linux Mint; compile fails with an error message indicating it can't find a file:
````
CMake Error at /home/kj/spack/opt/spack/linux-ubuntu20.10-westmere/gcc-10.2.0/qt-5.15.2-up6rrgb42pwybyn2bntm3q6itsagcdc2/lib/cmake/Qt5/Qt5Config.cmake:28 (find_package):
  Could not find a package configuration file provided by "Qt5OpenGL" with
  any of the following names:

    Qt5OpenGLConfig.cmake
    qt5opengl-config.cmake
````
The particular files that can't be found are different than with Linux Mint.  I'm not sure if that is meaningful or not.

Kim

**Reply by kimjohn1** [reply]
I figured out how to switch to the boost172 branch.  I still get the Cmake error.  The complete error message is:
````
CMake Deprecation Warning at CMakeLists.txt:30 (CMAKE_POLICY):
  The OLD behavior for policy CMP0043 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


CMake Error at /home/kj/Qt/5.15.2/gcc_64/lib/cmake/Qt5Gui/Qt5GuiConfigExtras.cmake:9 (message):
  Failed to find "GL/gl.h" in "/usr/include/libdrm".
Call Stack (most recent call first):
  /home/kj/Qt/5.15.2/gcc_64/lib/cmake/Qt5Gui/Qt5GuiConfig.cmake:232 (include)
  /home/kj/Qt/5.15.2/gcc_64/lib/cmake/Qt5/Qt5Config.cmake:28 (find_package)
  Superbuild.cmake:128 (FIND_PACKAGE)
  CMakeLists.txt:38 (INCLUDE)


-- Configuring incomplete, errors occurred!
See also "/home/kj/SCIRun/bin/CMakeFiles/CMakeOutput.log".
````
The CMakeOutput.log is attached.

Kim
[CMakeOutput.log](https://github.com/SCIInstitute/SCIRun/files/6025094/CMakeOutput.log)

**Reply by kimjohn1** [reply]
One more bit of info:
The error message indicates that the installer can't find "GL/gl.h" in "/usr/include/libdrm".  I note that there is no /usr/include/libdrm but there is a /usr/share/libdrm file that contains a list of AMD GPUs.  No help there.  I found neither a GL folder nor gl.h file anywhere (using the original and using the boost172 branch).

Kim

**Reply by kimjohn1** [reply]
The CMake error was the result of a missing dependency.  I installed libgl1-mesa-dev, and
````
cmake ../Superbuild
````
completes successfully.  I proceeded to the build step; sudo ./build.sh, which ran for about 30 minutes, and then produced an error message:
````
make[5]: *** No rule to make target '/home/kj/SCIRun-5.0-beta.Y/bin/Externals/Source/Boost_external/lib/libboost_python-mt.a', needed by 'lib/libCore_Utils.so'.  Stop.
make[4]: *** [CMakeFiles/Makefile2:5351: Core/Utils/CMakeFiles/Core_Utils.dir/all] Error 2
make[3]: *** [Makefile:152: all] Error 2
make[2]: *** [CMakeFiles/SCIRun_external.dir/build.make:126: download/FwdInvToolkit/Stamp/SCIRun_external/SCIRun_external-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:121: CMakeFiles/SCIRun_external.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
Building SCIRun returned an error

Either the code failed to build properly or

the testing programs failed to complete without

every single test program passing the test.
````
I believe I was using the boost 172 branch, though I'm not certain of that.  Let me know if you have any ideas on what I might do to resolve this error.

Kim

**Reply by dcwhite** [reply]
This is progress, but might be an error specifically on your system. If you switch to master branch and fetch it, reconfigure, and build again, and this still happens, there might be something weird going on with Mint. The build works on various Ubuntu versions, but I haven't tested every Linux distro of course.

**Reply by kimjohn1** [reply]
As long as we are having progress, we will eventually get there!

No luck with the different OS/different systems.  Here is what I have done today:

- Installed Linux Mint 20.10 on a lab computer and carefully performed a SCIRun Build in accordance with the Build instructions.
- Installed Ubuntu 20.04 on a different lab computer and did the same controlled SCIRun Build instructions.

Identical errors were reported in both cases (see the attached error reports)
[Linux_Mint_20.10_error_report.txt](https://github.com/SCIInstitute/SCIRun/files/6052820/Linux_Mint_20.10_error_report.txt)
[Ubuntu_20.04_error_report.txt](https://github.com/SCIInstitute/SCIRun/files/6052826/Ubuntu_20.04_error_report.txt)

I notice that the Build instructions that follow the cmake ../Superbuild step are non-specific:

>After configuration is done, generate the make files or project files for your favorite development environment and build.  Following the previous example, the SCIRun application will be built in bin/SCIRun.

Lacking any more imaginative options, I just used a simple make command there, and that command is what produced the reported errors.  Is there some other command or series of commands that I should be using after the cmake ../Superbuild step completes?

Kim

**Reply by kimjohn1** [reply]
I forgot to note that I used the master branch (and not the boost 172 branch) for the installation attempts today.

Kim

**Comment by kimjohn1** [comment, marked as answer]
Issue has been resolved.
